helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java

```diff
@@ -41,18 +60,279 @@
 public class WagedRebalancer implements GlobalRebalancer<ResourceControllerDataProvider> {
   private static final Logger LOG = LoggerFactory.getLogger(WagedRebalancer.class);
 
+  // When any of the following change happens, the rebalancer needs to do a global rebalance.
+  private static final Set<HelixConstants.ChangeType> GLOBAL_REBALANCE_REQUIRED_CHANGE_TYPES =
+      Collections.unmodifiableSet(new HashSet<>(Arrays
+          .asList(HelixConstants.ChangeType.RESOURCE_CONFIG, HelixConstants.ChangeType.CONFIG,
+              HelixConstants.ChangeType.INSTANCE_CONFIG)));
+
+  private ResourceChangeDetector _clusterDataDetector;
+  private AssignmentMetadataStore _assignmentMetadataStore;
+  private RebalanceAlgorithm _rebalanceAlgorithm;
+  private MappingCalculator<ResourceControllerDataProvider> _mappingCalculator;
+
   @Override
   public void init() {
```

Comment:
I don't think the init method is necessary. If the class is stateless we can create singleton instance. A constructor will give you the ability to mock the dependency for unit testing